### ISSUE TITLE
[release-1.23] New features from K3s: user-provided CA certs and `kubeadm` bootstrap token support

### ISSUE THREAD

**Issue by brandond** [rancher/rke2#3896]
RKE2 tracking issue for new K3s features:
* https://github.com/k3s-io/k3s/pull/6615
* https://github.com/k3s-io/k3s/pull/6663

See K3s PRs for functionality to test

#### User-Facing Change ####
```release-note
RKE2 now functions properly when the cluster CA certificates are signed by an existing root or intermediate CA. You can find a sample script for generating such certificates before RKE2 starts in the K3s repo at [contrib/util/certs.sh](https://github.com/k3s-io/k3s/blob/master/contrib/util/certs.sh).
RKE2 now supports `kubeadm` style join tokens. `rke2 token create` now creates join token secrets, optionally with a limited TTL.
RKE2 agents joined with an expired or deleted token stay in the cluster using existing client certificates via the NodeAuthorization admission plugin, unless their Node object is deleted from the cluster.
```

**Comment by ShylajaDevadiga** [comment]
### Validated on commit id b902ac88e84152a60d7ba3dde450de60d2f54da9 on release-1.23 branch
#### Environment Details

*Infrastructure*
Cloud EC2 instance

*Node(s) CPU architecture, OS, and Version:*
Ubuntu 20.04

*Cluster Configuration:*
3 server 1 agent

  *Config.yaml:*
```
cat <<EOF> /etc/rancher/rke2/config.yaml
write-kubeconfig-mode: 644
token: <TOKEN>
```
## Steps to validate:
1. Copy config.yaml 
3. Use certs.sh to generate CA certs before startup, note that cluster works properly and uses generated certs
```
$ curl -sfL https://get.rke2.io | sudo INSTALL_RKE2_COMMIT=b902ac88e84152a60d7ba3dde450de60d2f54da9 sh -
$ wget https://raw.githubusercontent.com/k3s-io/k3s/master/contrib/util/certs.sh
$ sudo -E PRODUCT=rke2 ./certs.sh
$ chmod u+x certs.sh 
$ sudo ./certs.sh 
$ sudo systemctl enable rke2-server
$ sudo systemctl start rke2-server
$ sudo ls -l /var/lib/rancher/rke2/server/tls
total 136
-rw-r--r-- 1 root root 1181 Feb 16 19:57 client-admin.crt
-rw------- 1 root root  227 Feb 16 19:57 client-admin.key
-rw-r--r-- 1 root root 1186 Feb 16 19:57 client-auth-proxy.crt
-rw------- 1 root root  227 Feb 16 19:57 client-auth-proxy.key
-rw-r--r-- 1 root root  574 Feb 16 19:57 client-ca.crt
-rw------- 1 root root  227 Feb 16 19:57 client-ca.key
-rw-r--r-- 1 root root  574 Feb 16 19:57 client-ca.nochain.crt
-rw-r--r-- 1 root root 1169 Feb 16 19:57 client-controller.crt
-rw------- 1 root root  227 Feb 16 19:57 client-controller.key
-rw-r--r-- 1 root root 1185 Feb 16 19:57 client-kube-apiserver.crt
-rw------- 1 root root  227 Feb 16 19:57 client-kube-apiserver.key
-rw-r--r-- 1 root root 1153 Feb 16 19:57 client-kube-proxy.crt
-rw------- 1 root root  227 Feb 16 19:57 client-kube-proxy.key
-rw------- 1 root root  227 Feb 16 19:57 client-kubelet.key
-rw-r--r-- 1 root root 1169 Feb 16 19:57 client-rke2-cloud-controller.crt
-rw------- 1 root root  227 Feb 16 19:57 client-rke2-cloud-controller.key
-rw-r--r-- 1 root root 1157 Feb 16 19:57 client-rke2-controller.crt
-rw------- 1 root root  227 Feb 16 19:57 client-rke2-controller.key
-rw-r--r-- 1 root root 1157 Feb 16 19:57 client-scheduler.crt
-rw------- 1 root root  227 Feb 16 19:57 client-scheduler.key
-rw-r--r-- 1 root root 4793 Feb 17 01:02 dynamic-cert.json
drwxr-xr-x 2 root root 4096 Feb 16 19:57 etcd
-rw-r--r-- 1 root root  595 Feb 16 19:57 request-header-ca.crt
-rw------- 1 root root  227 Feb 16 19:57 request-header-ca.key
-rw-r--r-- 1 root root  570 Feb 16 19:57 server-ca.crt
-rw------- 1 root root  227 Feb 16 19:57 server-ca.key
-rw-r--r-- 1 root root  570 Feb 16 19:57 server-ca.nochain.crt
-rw------- 1 root root 1675 Feb 16 19:57 service.current.key
-rw------- 1 root root 1675 Feb 16 19:57 service.key
-rw-r--r-- 1 root root 1380 Feb 16 19:57 serving-kube-apiserver.crt
-rw------- 1 root root  227 Feb 16 19:57 serving-kube-apiserver.key
-rw------- 1 root root  227 Feb 16 19:57 serving-kubelet.key
drwx------ 2 root root 4096 Feb 16 19:57 temporary-certs
```
Join other nodes nodes, confirmed that they are the same
```
curl -sfL https://get.rke2.io | sudo INSTALL_RKE2_COMMIT=b902ac88e84152a60d7ba3dde450de60d2f54da9 sh -
$ sudo ls -l /var/lib/rancher/rke2/server/tls
total 136
-rw-r--r-- 1 root root 1181 Feb 17 00:58 client-admin.crt
-rw------- 1 root root  227 Feb 17 00:58 client-admin.key
-rw-r--r-- 1 root root 1186 Feb 17 00:58 client-auth-proxy.crt
-rw------- 1 root root  227 Feb 17 00:58 client-auth-proxy.key
-rw------- 1 root root  574 Feb 16 19:57 client-ca.crt
-rw------- 1 root root  227 Feb 16 19:57 client-ca.key
-rw-r--r-- 1 root root  574 Feb 17 00:58 client-ca.nochain.crt
-rw-r--r-- 1 root root 1169 Feb 17 00:58 client-controller.crt
-rw------- 1 root root  227 Feb 17 00:58 client-controller.key
-rw-r--r-- 1 root root 1185 Feb 17 00:58 client-kube-apiserver.crt
-rw------- 1 root root  227 Feb 17 00:58 client-kube-apiserver.key
-rw-r--r-- 1 root root 1153 Feb 17 00:58 client-kube-proxy.crt
-rw------- 1 root root  227 Feb 17 00:58 client-kube-proxy.key
-rw------- 1 root root  227 Feb 17 00:58 client-kubelet.key
-rw-r--r-- 1 root root 1169 Feb 17 00:58 client-rke2-cloud-controller.crt
-rw------- 1 root root  227 Feb 17 00:58 client-rke2-cloud-controller.key
-rw-r--r-- 1 root root 1161 Feb 17 00:58 client-rke2-controller.crt
-rw------- 1 root root  227 Feb 17 00:58 client-rke2-controller.key
-rw-r--r-- 1 root root 1157 Feb 17 00:58 client-scheduler.crt
-rw------- 1 root root  227 Feb 17 00:58 client-scheduler.key
-rw-r--r-- 1 root root 4793 Feb 17 01:02 dynamic-cert.json
drwx------ 2 root root 4096 Feb 17 00:58 etcd
-rw------- 1 root root  595 Feb 16 19:57 request-header-ca.crt
-rw------- 1 root root  227 Feb 16 19:57 request-header-ca.key
-rw------- 1 root root  570 Feb 16 19:57 server-ca.crt
-rw------- 1 root root  227 Feb 16 19:57 server-ca.key
-rw-r--r-- 1 root root  570 Feb 17 00:58 server-ca.nochain.crt
-rw------- 1 root root 1675 Feb 17 00:58 service.current.key
-rw------- 1 root root 1675 Feb 16 19:57 service.key
-rw-r--r-- 1 root root 1380 Feb 17 00:58 serving-kube-apiserver.crt
-rw------- 1 root root  227 Feb 17 00:58 serving-kube-apiserver.key
-rw------- 1 root root  227 Feb 17 00:58 serving-kubelet.key
drwx------ 2 root root 4096 Feb 17 00:58 temporary-certs
```
```
rke2 version v1.23.16-dev+b902ac88 (b902ac88e84152a60d7ba3dde450de60d2f54da9)
go version go1.19.5 X:boringcrypto
NAME               STATUS   ROLES                       AGE     VERSION
ip-172-31-1-91     Ready    <none>                      4h37m   v1.23.16+rke2r1
ip-172-31-10-216   Ready    control-plane,etcd,master   9h      v1.23.16+rke2r1
ip-172-31-11-201   Ready    control-plane,etcd,master   4h46m   v1.23.16+rke2r1
ip-172-31-2-34     Ready    control-plane,etcd,master   4h44m   v1.23.16+rke2r1
NAMESPACE     NAME                                                    READY   STATUS      RESTARTS   AGE
kube-system   cloud-controller-manager-ip-172-31-10-216               1/1     Running     0          9h
kube-system   cloud-controller-manager-ip-172-31-11-201               1/1     Running     0          4h46m
kube-system   cloud-controller-manager-ip-172-31-2-34                 1/1     Running     0          4h44m
kube-system   etcd-ip-172-31-10-216                                   1/1     Running     0          9h
kube-system   etcd-ip-172-31-11-201                                   1/1     Running     0          4h46m
kube-system   etcd-ip-172-31-2-34                                     1/1     Running     0          4h44m
kube-system   helm-install-rke2-canal-nm4pd                           0/1     Completed   0          9h
kube-system   helm-install-rke2-coredns-fkcwn                         0/1     Completed   0          9h
kube-system   helm-install-rke2-ingress-nginx-p6qgk                   0/1     Completed   0          9h
kube-system   helm-install-rke2-metrics-server-flxkw                  0/1     Completed   0          9h
kube-system   kube-apiserver-ip-172-31-10-216                         1/1     Running     0          9h
kube-system   kube-apiserver-ip-172-31-11-201                         1/1     Running     0          4h45m
kube-system   kube-apiserver-ip-172-31-2-34                           1/1     Running     0          4h43m
kube-system   kube-controller-manager-ip-172-31-10-216                1/1     Running     0          9h
kube-system   kube-controller-manager-ip-172-31-11-201                1/1     Running     0          4h46m
kube-system   kube-controller-manager-ip-172-31-2-34                  1/1     Running     0          4h44m
kube-system   kube-proxy-ip-172-31-1-91                               1/1     Running     0          4h37m
kube-system   kube-proxy-ip-172-31-10-216                             1/1     Running     0          9h
kube-system   kube-proxy-ip-172-31-11-201                             1/1     Running     0          4h46m
kube-system   kube-proxy-ip-172-31-2-34                               1/1     Running     0          4h44m
kube-system   kube-scheduler-ip-172-31-10-216                         1/1     Running     0          9h
kube-system   kube-scheduler-ip-172-31-11-201                         1/1     Running     0          4h46m
kube-system   kube-scheduler-ip-172-31-2-34                           1/1     Running     0          4h44m
kube-system   rke2-canal-742r5                                        2/2     Running     0          4h37m
kube-system   rke2-canal-7blqb                                        2/2     Running     0          4h44m
kube-system   rke2-canal-cm48n                                        2/2     Running     0          4h46m
kube-system   rke2-canal-x5lbg                                        2/2     Running     0          9h
kube-system   rke2-coredns-rke2-coredns-775c5b4bb4-44cds              1/1     Running     0          9h
kube-system   rke2-coredns-rke2-coredns-775c5b4bb4-7ggrp              1/1     Running     0          4h46m
kube-system   rke2-coredns-rke2-coredns-autoscaler-695fc554c9-5rth9   1/1     Running     0          9h
kube-system   rke2-ingress-nginx-controller-blccn                     1/1     Running     0          4h37m
kube-system   rke2-ingress-nginx-controller-gjbgh                     1/1     Running     0          4h43m
kube-system   rke2-ingress-nginx-controller-kwch6                     1/1     Running     0          9h
kube-system   rke2-ingress-nginx-controller-vsqrw                     1/1     Running     0          4h45m
kube-system   rke2-metrics-server-644f588b5-lbk2t                     1/1     Running     0          9h
```

**Comment by ShylajaDevadiga** [comment]
## Issue: Add kubeadm style bootstrap token secret support
While validating using commit on 09add0b982a978f1d5afddffad587f20431e2e5c on release-1.23 branch, found agent is unable to join the server. 

Issue is also found on other versions while validating latest commit

**Infrastructure**
Cloud EC2 instance

**Node(s) CPU architecture, OS, and Version:**
Ubuntu 20.04

**Cluster Configuration:**
3 server 1 agent

**Steps to reproduce**

Server
```
kubectl get nodes
ip-172-31-4-56   Ready   etcd    10s   v1.26.1+rke2r1
```
```
$ sudo rke2 token create
```
Agent
```
cat /etc/rancher/rke2/config.yaml 
token: <token generated on the server>
server: https://<SERVERIP>:9345
```
Install agent: 
```
curl -sfL https://get.rke2.io | sudo INSTALL_RKE2_COMMIT=09add0b982a978f1d5afddffad587f20431e2e5c sh -
```
Agent logs:
```
Feb 17 09:08:15 ip-172-31-1-83 rke2[1636]: time="2023-02-17T09:08:15Z" level=info msg="Waiting to retrieve agent configuration; server is not ready: failed to retrieve configuration from server: https://127.0.0.1:6444/v1-rke2/config: 403 Forbidden"
Feb 17 09:08:20 ip-172-31-1-83 rke2[1636]: time="2023-02-17T09:08:20Z" level=info msg="Waiting to retrieve agent configuration; server is not ready: failed to retrieve configuration from server: https://127.0.0.1:6444/v1-rke2/config: 403 Forbidden"
```

**Comment by ShylajaDevadiga** [comment]
Validated fix on release-1.23 branch using commit id af7a5edbc24ab5e907bc4619a8b3261b84b24237

- Joined agent successfully
- Rebooted agent node
- Rebooted all nodes
- Restart agent service

```
ubuntu@ip-172-31-15-86:~$ kubectl get nodes
NAME              STATUS   ROLES                       AGE     VERSION
ip-172-31-0-51    Ready    <none>                      10m     v1.23.16+rke2r1
ip-172-31-11-34   Ready    control-plane,etcd,master   9m59s   v1.23.16+rke2r1
ip-172-31-15-86   Ready    control-plane,etcd,master   61m     v1.23.16+rke2r1
ip-172-31-3-23    Ready    <none>                      3m44s   v1.23.16+rke2r1
ip-172-31-9-64    Ready    control-plane,etcd,master   10m     v1.23.16+rke2r1
ubuntu@ip-172-31-15-86:~$ kubectl get pods -A 
NAMESPACE     NAME                                                    READY   STATUS      RESTARTS   AGE
kube-system   cloud-controller-manager-ip-172-31-11-34                1/1     Running     0          9m49s
kube-system   cloud-controller-manager-ip-172-31-15-86                1/1     Running     0          61m
kube-system   cloud-controller-manager-ip-172-31-9-64                 1/1     Running     0          9m46s
kube-system   etcd-ip-172-31-11-34                                    1/1     Running     0          9m52s
kube-system   etcd-ip-172-31-15-86                                    1/1     Running     0          61m
kube-system   etcd-ip-172-31-9-64                                     1/1     Running     0          9m48s
kube-system   helm-install-rke2-canal-7sl6w                           0/1     Completed   0          61m
kube-system   helm-install-rke2-coredns-rqcwk                         0/1     Completed   0          61m
kube-system   helm-install-rke2-ingress-nginx-5xknq                   0/1     Completed   0          61m
kube-system   helm-install-rke2-metrics-server-rw6js                  0/1     Completed   0          61m
kube-system   kube-apiserver-ip-172-31-11-34                          1/1     Running     0          9m33s
kube-system   kube-apiserver-ip-172-31-15-86                          1/1     Running     0          61m
kube-system   kube-apiserver-ip-172-31-9-64                           1/1     Running     0          10m
kube-system   kube-controller-manager-ip-172-31-11-34                 1/1     Running     0          9m57s
kube-system   kube-controller-manager-ip-172-31-15-86                 1/1     Running     0          61m
kube-system   kube-controller-manager-ip-172-31-9-64                  1/1     Running     0          9m47s
kube-system   kube-proxy-ip-172-31-0-51                               1/1     Running     0          10m
kube-system   kube-proxy-ip-172-31-11-34                              1/1     Running     0          9m54s
kube-system   kube-proxy-ip-172-31-15-86                              1/1     Running     0          61m
kube-system   kube-proxy-ip-172-31-3-23                               1/1     Running     0          3m45s
kube-system   kube-proxy-ip-172-31-9-64                               1/1     Running     0          9m48s
kube-system   kube-scheduler-ip-172-31-11-34                          1/1     Running     0          9m57s
kube-system   kube-scheduler-ip-172-31-15-86                          1/1     Running     0          61m
kube-system   kube-scheduler-ip-172-31-9-64                           1/1     Running     0          9m53s
kube-system   rke2-canal-4tdg8                                        2/2     Running     0          10m
kube-system   rke2-canal-cnjzd                                        2/2     Running     0          61m
kube-system   rke2-canal-hjjjj                                        2/2     Running     0          10m
kube-system   rke2-canal-wwj9n                                        2/2     Running     0          10m
kube-system   rke2-canal-xlsb2                                        2/2     Running     0          3m46s
kube-system   rke2-coredns-rke2-coredns-775c5b4bb4-fnfhm              1/1     Running     0          10m
kube-system   rke2-coredns-rke2-coredns-775c5b4bb4-h8bdl              1/1     Running     0          61m
kube-system   rke2-coredns-rke2-coredns-autoscaler-695fc554c9-vjb7l   1/1     Running     0          61m
kube-system   rke2-ingress-nginx-controller-77bw9                     1/1     Running     0          9m31s
kube-system   rke2-ingress-nginx-controller-fxrrs                     1/1     Running     0          9m30s
kube-system   rke2-ingress-nginx-controller-hg4cx                     1/1     Running     0          10m
kube-system   rke2-ingress-nginx-controller-lzjwr                     1/1     Running     0          60m
kube-system   rke2-ingress-nginx-controller-rh5x7                     1/1     Running     0          2m45s
kube-system   rke2-metrics-server-644f588b5-w7gjp                     1/1     Running     0          60m
```
